toolchain/check/scope_stack.cpp

```diff
@@ -142,4 +139,34 @@ auto ScopeStack::SetReturnedVarOrGetExisting(SemIR::InstId inst_id)
   return SemIR::InstId::Invalid;
 }
 
+auto ScopeStack::Suspend() -> SuspendedScope {
+  CARBON_CHECK(!scope_stack_.empty()) << "No scope to suspend";
+  SuspendedScope result = {scope_stack_.pop_back_val(), {}};
+  if (result.entry.scope_id.is_valid()) {
+    non_lexical_scope_stack_.pop_back();
+  }
+  for (auto name_id : result.entry.names) {
+    result.suspended_lookups.push_back(lexical_lookup_.Suspend(name_id));
+  }
+  // This would be easy to support if we had a need, but currently we do not.
+  CARBON_CHECK(!result.entry.has_returned_var)
+      << "Should not suspend a scope with a returned var.";
+  return result;
+}
+
+auto ScopeStack::Restore(SuspendedScope scope) -> void {
+  for (auto entry : scope.suspended_lookups) {
+    // clang-tidy warns that the `std::move` below has no effect. While that's
+    // true, this `move` defends against the suspended lookup growing more state
+    // later.
```

Comment:
I'm a little confused by this, the parameter is taken by value so I wouldn't expect it to be that defensive.

Reply:
It's taken by value, but we're moving into the function parameter. If the suspension object owned something (eg, a tiny vector), this'd mean we move that thing rather than copying it.

The clang-tidy check is just bogus in this situation: it's preventing us from expressing a semantic move because it's looked at the implementation details of the type and determined that it would happen to be the same as a copy, but that is not a reasonable reason to issue a complaint -- it's expecting us to rely on implementation details of a type we don't own, and not express what we mean.

However, I think the check is probably what we want in the majority of cases, and sadly the same clang-tidy check name also covers other situations where the check is detecting a bug (moving a const object), so I don't think we should turn it off.